tensorflow_probability/python/math/special.py

```diff
@@ -88,280 +88,378 @@ def atan_difference(x, y, name=None):
+
+  # Remove the previously added extra dimension: it is no longer needed.
+  f = tf.squeeze(values[-1], axis=-1)
+  f_grad_a, f_grad_b = tf.unstack(gradients[-1], axis=-1)
+
+  return f, f_grad_a, f_grad_b
+
+
+def _betainc_der_continued_fraction(a, b, x, dtype, where):
+  """Returns the partial derivatives of betainc with respect to a and b."""
+  # This function evaluates betainc(a, b, x) by its continued fraction
+  # expansion given here: https://dlmf.nist.gov/8.17.E22
+  # We apply this function when the input (a, b, x) does not belong to the
+  # proper region of computation of `_betainc_der_power_series`.
+  one = tf.constant(1., dtype=dtype)
+  two = tf.constant(2., dtype=dtype)
+
+  # This continued fraction expansion of betainc converges rapidly
+  # for x < (a - 1) / (a + b - 2). For x >= (a - 1) / (a + b - 2),
+  # we can obtain an equivalent computation by using the symmetry
+  # relation given here: https://dlmf.nist.gov/8.17#E4
+  apply_symmetry = (x >= (a - one) / (a + b - two))
+  a_orig = a
+  a = tf.where(apply_symmetry, b, a)
+  b = tf.where(apply_symmetry, a_orig, b)
+  x = tf.where(apply_symmetry, one - x, x)
+
+  cf, cf_grad_a, cf_grad_b = _betainc_modified_lentz_method(
+      a, b, x, dtype, where)
+
+  more_terms = tf.math.exp(
+      tf.math.xlogy(a, x) + tf.math.xlog1py(b, -x) -
+      tf.math.log(a) - lbeta(a, b))
 
+  digamma_apb = tf.math.digamma(a + b)
+  grad_a = more_terms * (cf_grad_a + cf * (tf.math.log(x) -
+      tf.math.reciprocal(a) + digamma_apb - tf.math.digamma(a)))
+  grad_b = more_terms * (cf_grad_b + cf * (tf.math.log1p(-x) +
+      digamma_apb - tf.math.digamma(b)))
+
+  # If we are taking advantage of the symmetry relation, then we have to
+  # adjust grad_a and grad_b.
+  grad_a_orig = grad_a
+  grad_a = tf.where(apply_symmetry, -grad_b, grad_a)
+  grad_b = tf.where(apply_symmetry, -grad_a_orig, grad_b)
+
+  return grad_a, grad_b
+
+
+def _betainc_der_power_series(a, b, x, dtype, where):
+  """Returns the partial derivatives of betainc with respect to a and b."""
+  # This function evaluates betainc(a, b, x) by its series representation:
+  #   x ** a * 2F1(a, 1 - b; a + 1; x) / (a * B(a, b)) ,
+  # where 2F1 is defined here: http://dlmf.nist.gov/15.2.i
+  # We apply this function when the input (a, b, x) satisfies at least one
+  # of the following conditions:
+  #   C1: (x < a / (a + b)) & (b * x <= 1) & (x <= 0.95)
+  #   C2: (x >= a / (a + b)) & (a * (1 - x) <= 1) & (x >= 0.05)
+  numpy_dtype = dtype_util.as_numpy_dtype(dtype)
+  eps = tf.constant(np.finfo(numpy_dtype).eps, dtype=dtype)
+  half = tf.constant(0.5, dtype=dtype)
+  one = tf.constant(1., dtype=dtype)
+
+  # Avoid returning NaN or infinity when the input does not satisfy either
+  # C1 or C2.
+  safe_a = tf.where(where, a, half)
+  safe_b = tf.where(where, b, half)
+  safe_x = tf.where(where, x, half)
+
+  # When the condition C1 is false, we apply the symmetry relation given
+  # here: http://dlmf.nist.gov/8.17.E4
```

Comment:
It would be nice (here and elsewhere) to inline the relationship, that way it is easier to see at a glance what the variable / code is referring to.

Reply:
I'm not sure if I understood this comment. Is it about making the symmetry relationship explicit in the code comment, instead of using a link?

Reply:
Sorry, yes I mean making it explicit in the code comment (and leaving the link in case someone wants to read more).

Reply:
Done.